### ISSUE TITLE
Intelligent mode does not find all solutions

### ISSUE THREAD

**Issue by viktoras-pal** [thomas-maeder/popeye#307]
begin
pieces
white Ka8 Sf3
black Rb7b1 Kh1
stipulation h#7
option intelligent
end

1.Rb1-f1 Sf3-h4 2.Rb7-b1 Ka8-a7 3.Rb1-a1+ Ka7-b6 4.Kh1-g1 Kb6-c5 5.Kg1-f2 Kc5-d4 6.Kf2-e1 Kd4-e3 7.Ra1-d1 Sh4-g2#
The following solution is missing:
1.Rb7-f7 Sf3-h4 2.Rf7-f1 Ka8-a7 3.Rb1-a1+ Ka7-b6 4.Kh1-g1 Kb6-c5 5.Kg1-f2 Kc5-d4 6.Kf2-e1 Kd4-e3 7.Ra1-d1 Sh4-g2#

**Comment by JoshuaGreen** [comment]
I see the same thing on my computer.  Thank you for bringing this to our attention.

**Comment by thomas-maeder** [comment]
There is a strange interaction between intelligent mode and the hashtable.

If the hashtable is suppressed with the command line option -maxmem 0, Popeye finds both solutions.

If the hashtable is active, solving stops after 1.Rb7-f7 Sf3-h4 2.Rf7-f1 after the resulting position is found in the hashtable, probably from 1.Rb1-f1 Sf3-h4 2.Rb7-b1. But why exactly solving stops isn't clear yet.

FWIW,
begin
pieces white Ka8 Sf3 black Rb7b1 Kh1
stipulation a=>b2
pieces white ka7 sh4 black rb1f1 kh1
end
finds both solutions.

More analysis is needed.

**Comment by JoshuaGreen** [comment]
A bit more evidence for now -- with `-maxmem 7` my Popeye finds both solutions but with `-maxmem 8` it does not.

**Comment by JoshuaGreen** [comment]
I've done a bit more digging.  With `-maxmem 7` and hacking [`optimisations/intelligent/intelligent.c`](https://github.com/thomas-maeder/popeye/blob/develop/optimisations/intelligent/intelligent.c) to _only_ try to solve the relevant target position, I find that Popeye considers the moves **1. Rf7 Sh4&nbsp; 2. fRf1 Ka7** but then doesn't take the line any further.&nbsp; I'm currently trying to determine why it gives up here.

**Comment by JoshuaGreen** [comment]
In the above run, it _appears_ that Popeye finds the position in the hash table with a `solve_result` of 15 and gives up since `solve_nr_remaining == 13` at that point.&nbsp; I don't understand the hash table well enough to figure out what's going on here.

**Comment by TorstenLinss** [comment]
Interestingly, version 4.37 finds both solutions in 0.020 sec, while 4.87 takes 1.04 sec [and finds only one].

**Comment by TorstenLinss** [comment]
One observation more: version 4.37 analyses 20 final position, 4.87 only 19.

I think, in my original code (in 1997) I treated two (and) more rooks as different (artificially making them different by setting DiaRenSomething). Then I sent Rooks a/b to squares x/y and for the second possibility to y/x. This somehow seems to got lost.

**Comment by JoshuaGreen** [comment]
I can't say that I really understand what the hash table is storing, but intelligent mode treats the Rooks differently (based on their starting positions; different permutations yield different target positions) so the hash table _in that mode_ must do the same.&nbsp; I'd imagine that issues would have turned up sooner if that weren't happening.

**Comment by viktoras-pal** [comment]
Perhaps this is due to a possible collision if two different positions have the same hash value.

**Comment by thomas-maeder** [comment]
Thanks for all the contributions, especially from Torsten who originally implemented intelligent mode.

What was primarily lost since 4.37 is the following:
```
    if (OptFlag[intelligent]) { /* V3.76  TLi */
         one_byte_hash = false;
         bytes_for_spec= 4;
    }

```
The rook's identities are then stored in these 4 bytes_for_spec.

**Comment by JoshuaGreen** [comment]
Phew, it wasn't something I did. :P

More seriously, good catch @thomas-maeder.&nbsp; I'm still a bit concerned that a bug like this could have been hiding for so long; how was it not noticed before?&nbsp; Composers may want to recheck their problems.

**Comment by JoshuaGreen** [comment]
@TorstenLinss, thank you for giving us intelligent mode.

**Comment by thomas-maeder** [comment]
Fixed for 4.87
